pulsar-function-go/pf/instance.go

```diff
@@ -0,0 +1,289 @@
+					Properties:        properties,
+					Type:              consumerType,
+					MessageChannel:    channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+				gi.consumers = append(gi.consumers, consumer)
+			} else {
+				consumer, err := gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:             topic,
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              consumerType,
+					ReceiverQueueSize: int(consumerConf.ReceiverQueueSize.Value),
+					MessageChannel:    channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+
+				gi.consumers = append(gi.consumers, consumer)
+			}
+		} else {
+			if consumerConf.IsRegexPattern {
+				consumer, err := gi.client.Subscribe(pulsar.ConsumerOptions{
+					TopicsPattern:    topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             consumerType,
+					MessageChannel:   channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+				gi.consumers = append(gi.consumers, consumer)
+			} else {
+				consumer, err := gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:            topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             consumerType,
+					MessageChannel:   channel,
+				})
+				if err != nil {
+					log.Errorf("create consumer error:%s", err.Error())
+				}
+				gi.consumers = append(gi.consumers, consumer)
+			}
+		}
+		gi.context.InputTopics = append(gi.context.InputTopics, topic)
+
+	CLOSE:
+		for {
+			select {
+			case cm := <-channel:
+				gi.msg = append(gi.msg, cm.Message)
+			case <-time.After(time.Millisecond * 500):
+				close(channel)
+				break CLOSE
+			}
+		}
+	}
+
+	gi.actualExecution(function)
+	gi.close()
+}
+
+func (gi *GoInstance) actualExecution(function Function) {
+	gi.function = function
+	output, err := gi.handlerMsg()
+	err = gi.processResult(output)
+	if err != nil {
+		log.Errorf("process output result err:%v", err)
+		return
+	}
+}
+
+func (gi *GoInstance) processResult(output []byte) error {
+	gi.context = NewFuncContext()
+	atLeastOnce := gi.context.InstanceConf.FuncDetails.ProcessingGuarantees == pb.ProcessingGuarantees_ATLEAST_ONCE
+
+	if output != nil && gi.context.InstanceConf.FuncDetails.Sink.Topic != "" &&
+		len(gi.context.InstanceConf.FuncDetails.Sink.Topic) > 0 {
+		if gi.producer == nil {
+			err := gi.setUpProducer()
+			if err != nil {
+				log.Errorf("set up producer error:%s", err.Error())
+				return err
+			}
+		}
+
+		asyncMsg := pulsar.ProducerMessage{
+			Payload: output,
+		}
+		// Attempt to send the message asynchronously and handle the response
+		gi.producer.SendAsync(context.Background(), asyncMsg, func(message pulsar.ProducerMessage, e error) {
+			if e != nil {
+				log.Fatal(e)
+			}
+		})
+
+		for _, consumer := range gi.consumers {
```

Comment:
you need a consumer map from topic-name to consumers (an array doesn't work efficiently here). so you can find the consumer by looking up the map using `input.Topic()`.

you can implement the ack logic in a function `ackInputMessage(input Message)`. so you can use this method when you need to ack the input message.